terminal-view/src/main/java/com/termux/view/TerminalView.java

```diff
@@ -0,0 +1,1128 @@
+        } else if (event.isSystem() && (!mClient.shouldBackButtonBeMappedToEscape() || keyCode != KeyEvent.KEYCODE_BACK)) {
+            return super.onKeyDown(keyCode, event);
+        } else if (event.getAction() == KeyEvent.ACTION_MULTIPLE && keyCode == KeyEvent.KEYCODE_UNKNOWN) {
+            mTermSession.write(event.getCharacters());
+            return true;
+        }
+
+        final int metaState = event.getMetaState();
+        final boolean controlDown = event.isCtrlPressed() || mClient.readControlKey();
+        final boolean leftAltDown = (metaState & KeyEvent.META_ALT_LEFT_ON) != 0 || mClient.readAltKey();
+        final boolean rightAltDownFromEvent = (metaState & KeyEvent.META_ALT_RIGHT_ON) != 0;
+
+        int keyMod = 0;
+        if (controlDown) keyMod |= KeyHandler.KEYMOD_CTRL;
+        if (event.isAltPressed() || leftAltDown) keyMod |= KeyHandler.KEYMOD_ALT;
+        if (event.isShiftPressed()) keyMod |= KeyHandler.KEYMOD_SHIFT;
+        if (event.isNumLockOn()) keyMod |= KeyHandler.KEYMOD_NUM_LOCK;
+        if (!event.isFunctionPressed() && handleKeyCode(keyCode, keyMod)) {
+            if (TERMINAL_VIEW_KEY_LOGGING_ENABLED) mClient.logInfo(LOG_TAG, "handleKeyCode() took key event");
+            return true;
+        }
+
+        // Clear Ctrl since we handle that ourselves:
+        int bitsToClear = KeyEvent.META_CTRL_MASK;
+        if (rightAltDownFromEvent) {
+            // Let right Alt/Alt Gr be used to compose characters.
+        } else {
+            // Use left alt to send to terminal (e.g. Left Alt+B to jump back a word), so remove:
+            bitsToClear |= KeyEvent.META_ALT_ON | KeyEvent.META_ALT_LEFT_ON;
+        }
+        int effectiveMetaState = event.getMetaState() & ~bitsToClear;
+
+        int result = event.getUnicodeChar(effectiveMetaState);
+        if (TERMINAL_VIEW_KEY_LOGGING_ENABLED)
+            mClient.logInfo(LOG_TAG, "KeyEvent#getUnicodeChar(" + effectiveMetaState + ") returned: " + result);
+        if (result == 0) {
+            return false;
+        }
+
+        int oldCombiningAccent = mCombiningAccent;
+        if ((result & KeyCharacterMap.COMBINING_ACCENT) != 0) {
+            // If entered combining accent previously, write it out:
+            if (mCombiningAccent != 0)
+                inputCodePoint(mCombiningAccent, controlDown, leftAltDown);
+            mCombiningAccent = result & KeyCharacterMap.COMBINING_ACCENT_MASK;
+        } else {
+            if (mCombiningAccent != 0) {
+                int combinedChar = KeyCharacterMap.getDeadChar(mCombiningAccent, result);
+                if (combinedChar > 0) result = combinedChar;
+                mCombiningAccent = 0;
+            }
+            inputCodePoint(result, controlDown, leftAltDown);
+        }
+
+        if (mCombiningAccent != oldCombiningAccent) invalidate();
+
+        return true;
+    }
+
+    public void inputCodePoint(int codePoint, boolean controlDownFromEvent, boolean leftAltDownFromEvent) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 27. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 678-696 into a separate method.